linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -683,10 +681,21 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
 
     val thisType = if (arrow) NoType else AnyType
 
-    val (allNewParamsWithUse, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
+    val (inlinableCaptures, normalCaptureParams) = {
+      val (i, n) = captureParams.iterator.zip(tcaptureValues.iterator).partition {
+        _._2 match {
+          // TODO: Add more criteria here
+          case PreTransLocalDef(LocalDef(_, _, _: ReplaceWithThis | _: ReplaceWithConstant)) => true
+          case _ => false
+        }
+      }
+      (i.toList, n.map(_._1).toList)
+    }
+
+    val bodyResult = transformIsolatedBody(None, thisType, inlinableCaptures.map(_._1),
+        normalCaptureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
```

Comment:
Instead of making the captures parameters here, consider re-creating bindings for the local defs. That will (IIUC) inline "more".